omigami/models/estimator.py

```diff
@@ -0,0 +1,105 @@
+import numpy as np
+from typing import Any
+from sklearn.base import BaseEstimator, clone
+from sklearn.pipeline import Pipeline
+from sklearn.ensemble import RandomForestClassifier
+from xgboost import XGBClassifier
+from omigami.data import RandomState
+from omigami.models.pls_classifier import PLSClassifier
+
+
+class ESTIMATORS:
+    RFC = "RFC"
+    XGBC = "XGBC"
+    PLSC = "PLSC"
+
+
+class Estimator:
+    def __init__(self, estimator: Any, random_state: RandomState):
+        self._estimator = estimator
+        self._random_state = random_state
+        self.set_random_state(random_state)
+
+    def fit(self, X, y):
+        self._estimator.fit(X, y)
+        return self
+
+    def predict(self, X):
+        return self._estimator.predict(X)
+
+    @property
+    def feature_importances(self):
+        raise NotImplementedError("feature_importance is not implemented")
+
+    def set_random_state(self, random_state: RandomState):
+        raise NotImplementedError("set_random_state is not implemented")
+
+    def _clone_estimator(self):
+        raise NotImplementedError("_clone_estimator is not implemented")
+
+    def clone(self):
+        estimator_clone = self._clone_estimator()
+        return self.__class__(estimator_clone, self._random_state)
+
+
+class ScikitLearnEstimator(Estimator):
+    @property
+    def feature_importances(self):
+        if hasattr(self._estimator, "feature_importances_"):
+            return self._estimator.feature_importances_
+        if hasattr(self._estimator, "coef_"):
+            return np.abs(self._estimator.coef_[0])
+        raise ValueError("The estimator provided has no feature importances")
+
+    def set_random_state(self, random_state: RandomState):
+        try:
+            self._estimator.set_params(random_state=random_state)
+        except ValueError:
+            pass  # not all models have a random_state param (e.g. LinearRegression)
+
+    def _clone_estimator(self):
+        return clone(self._estimator)
+
+
+class ScikitLearnPipeline(Estimator):
+    @property
+    def feature_importances(self):
+        for _, step in self._estimator.steps:
```

Comment:
Maybe instead of searching for it, the user could specify which step contains the feature importances in the constructor?

Reply:
I don't know, this would require that the user calls the wrapper himself. So far this is hot it works:
```python
pipeline = sklearn.pipeline.Pipeline([
    ("log", FunctionTransformer(np.log1p)),
    ("norm", Normalizer()),
    ("model", PLSClassifier())
])
fs = FeatureSelector(n_outer=6, n_repetitions=15, metric="MISS", estimator=pipeline)
```